Fix issue 13854 --  Appending to an interior slice of a large array results in unnecessary 16-byte offset

schveiguy opened the pull request:
When working with an existing block that is PAGESIZE or more, there was an optimization to always assume LARGEPAD and that we needed the offset. However, if the slice appended to was being reallocated to prevent stomping, and the new size fits into a smaller block, then the 16-byte offset is not only not required, but puts the array size in the wrong place.
This new code pushes all code paths through the same spot where the allocation and update to the new block is done correctly based on the new size.

MartinNowak commented:
This code seems to have caused a performance regression.
The [slist](https://github.com/D-Programming-Language/druntime/blob/master/benchmark/gcbench/slist.d) GC benchmark now runs about 50% slower, because almost twice as many collections are triggered.

schveiguy commented:
Without having looked at anything, this makes almost no sense. What is slist GC benchmark doing with array appending? And with arrays that are PAGE size?

MartinNowak commented:
> Without having looked at anything, this makes almost no sense. What is slist GC benchmark doing with array appending? And with arrays that are PAGE size?

Maybe it's the [`text.split()`](https://github.com/D-Programming-Language/druntime/blob/841c9a68cc209aa2abcab95593dce5f9e777b4c3/benchmark/gcbench/slist.d#L33) call, in any case, please find out.

MartinNowak commented:
Looks like before this change the words array of split could be extended in place, while after this change it's quite often needs a reallocation in between.

schveiguy commented:
It doesn't. It's not this patch. I switched it back to the code before, and it has the same problem. Still trying to investigate.

schveiguy commented:
The issue is not in lifetime.d

I reverted the file back to the 2.066.1 version, fixed one issue in API difference (when allocating a struct, now the struct typeinfo is passed instead of the pointer info), and then it still has the same issue.

Something else changed in the runtime to create this behavior.

schveiguy commented:
BTW, for reference, I instrumented split to see any pattern of allocation. I'm printing out when it reallocates, and how many elements in the array. The results are interesting, it appears that the faster code reallocates MORE often.

With latest head, pattern looks like:

```
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 16383
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 16383
```

With 2.066.1, pattern looks like:

```changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 30719
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 27391
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 30719
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 30719
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 25599
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 30719
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 27391
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
changed pointers, 30719
changed pointers, 1
changed pointers, 2
changed pointers, 4
changed pointers, 8
changed pointers, 16
changed pointers, 32
changed pointers, 64
changed pointers, 128
```
Note here, 128 elements = 128 * 16 bytes == 2048. This is the point at which extend is used. So any time you see no other number after 128, it was able to extend all the way.

I notice the number at which extend stops working is also much higher, which is counter-intuitive to me. A higher number means more copying, I would attribute that with less performance.

It appears to me that in the latest code, every other run results in extending all the way through. But on 2.066.1, extending only works at most 25% of the time all the way through. But it's noticeably faster, as you say.

MartinNowak commented:
Thanks for looking into this. The GC should be deterministic, and I can see different GC behavior before and after this pull request was merged.
So the only question is, how does this pull change qalloc/extend calls or their order?

schveiguy commented on src/rt/lifetime.d:
```diff
@@ -1823,15 +1819,7 @@ byte[] _d_arrayappendcTX(const TypeInfo ti, ref byte[] px, size_t n)
-                memcpy(newdata, px.ptr, length * sizeelem);
-                // do postblit processing
-                __doPostblit(newdata, length * sizeelem, ti.next);
-                (cast(void **)(&px))[1] = newdata;
```
This is the original code that runs during the `split` operation. I verified by `printf`during the test. After the pull, it's a goto to L2. Compare with that code. It's identical except for using info to determine whether to do LARGEPAD or assuming LARGEPAD, and doing a prefix, or assuming a prefix. everything else is identical.

In the case of slist test, the case that this code was intended to fix doesn't occur (appending to a small slice of a PAGE sized+ block), so the result should also be identical. I've looked at this code about 10 times, there is no difference I can see. Something other than this pull has to be the cause.

MartinNowak replied:
OK, found it.
The code slightly changed the stack layout, thereby getting rid of a false pointer that pinned one of the big words array.
Therefor the old code grew the heap (used an additional pool) and triggered less collections.

schveiguy replied:
Ugh.. The precise GC cannot come soon enough...
